app-2/src/main/java/tech/relaycorp/letro/awala/AwalaManager.kt

```diff
@@ -0,0 +1,235 @@
+        } else {
+            isAwalaInstalledOnDevice ?: checkIfAwalaAppInstalled()
+        }
+    }
+
+    private suspend fun startReceivingMessages() {
+        if (isReceivingMessages) {
+            return
+        }
+
+        awalaScope.launch {
+            Log.d(TAG, "start receiving messages...")
+            GatewayClient.receiveMessages().collect { message ->
+                // TODO Remove first message.ack() before publishing the app.
+                // It's here to avoid the server getting stuck with messages that can't be processed.
+                message.ack()
+                _messages.emit(
+                    Message(
+                        type = MessageType.from(message.type),
+                        content = message.content,
+                    ).also { Log.d(TAG, "Receive message: ($it)") }
+                )
+                message.ack()
+            }
+        }
+
+        isReceivingMessages = true
+    }
+
+    private suspend fun configureAwala() {
+        registerFirstPartyEndpointIfNeeded()
+        importServerThirdPartyEndpointIfNeeded()
+        authoriseReceivingMessagesFromThirdPartyEndpointIfNeeded()
+        startReceivingMessages()
+    }
+
+    private suspend fun checkIfAwalaAppInstalled(): Boolean {
+        try {
+            GatewayClient.bind()
+            configureAwala()
+        } catch (exp: GatewayBindingException) {
+            this.isAwalaInstalledOnDevice = false
+            return false
+        }
+        this.isAwalaInstalledOnDevice = true
+        return true
+    }
+
+    private suspend fun registerFirstPartyEndpointIfNeeded(): FirstPartyEndpoint? {
+        if (awalaRepository.getServerFirstPartyEndpointNodeId() != null) {
+            return null
+        }
+        val firstPartyEndpoint = FirstPartyEndpoint.register()
+        awalaRepository.saveServerFirstPartyEndpointNodeId(firstPartyEndpoint.nodeId)
+        return firstPartyEndpoint
+    }
+
+    private suspend fun authoriseReceivingMessagesFromThirdPartyEndpointIfNeeded() {
+        if (awalaRepository.isAuthorizedReceiveMessagesFromServer()) {
+            return
+        }
+        val firstPartyEndpointNodeId = awalaRepository.getServerFirstPartyEndpointNodeId()
+            ?: registerFirstPartyEndpointIfNeeded()?.nodeId
+            ?: ""
+
+        val thirdPartyEndpointNodeId = awalaRepository.getServerThirdPartyEndpointNodeId()
+            ?: importServerThirdPartyEndpointIfNeeded()?.nodeId
+            ?: ""
+
+        if (firstPartyEndpointNodeId.isEmpty() || thirdPartyEndpointNodeId.isEmpty()) {
+            // TODO: log error
+            Log.i(TAG, "authoriseReceivingMessagesFromThirdPartyEndpoint(). Some of ids is empty: $firstPartyEndpointNodeId ; $thirdPartyEndpointNodeId")
+            return
+        }
+
+        val firstPartyEndpoint = loadNonNullFirstPartyEndpoint(firstPartyEndpointNodeId)
+        val thirdPartyEndpoint = loadNonNullThirdPartyEndpoint(thirdPartyEndpointNodeId)
```

Comment:
Shouldn't these lines be replaced with `sendMessage()`?

Reply:
Yeap, I forgot to refactor it